### ISSUE TITLE
Stack<T> should implement IReadOnlyList<T>

### ISSUE THREAD

**Issue by michaelstaib** [dotnet/corefx#38596]
Hi,

it would be nice if `Stack<T>` also implemented `IReadOnlyList<T>`. 

`IReadOnlyList<T>` would allow to pick any element of the stack with its index. I know that the current implementation is a pretty clean stack implementation, but sometimes it is nice to just be able to pick an element of that stack.

With the current implementation `Stack<T>` already is implementing IReadOnlyCollection<T> which allows to iterate over it and retrieve the size of that collection. 

`IReadOnlyList<T>` would make implementing some algorithms more efficient while still allowing to use `Stack<T>`.

In my case I have reimplemented the stack in order to have that indexer.

**Comment by khellang** [comment]
Isn't the fact that `Stack<T>` allows random access internally just an implementation detail?

**Comment by michaelstaib** [comment]
I think I have to clarify what I want here since I see the negative reactions to my proposal.

So, what I want here is to be able to peek not only at the next item on the stack but alsoat one of my choosing without needing to pop an element. Since `Stack<T>` uses  an array underneath this would be pretty efficient.

Also it would allow me to iterate over the stack more efficiently.

The `Stack<T>` implementation already implements `IReadOnlyCollection<T>` and `ICollection` meaning we could also peek at things that are not on the top by using skip/take, my proposal is just to allow peeking at a specific entry in a more efficient way with an index.

We could also add a specific function here that would allow for a peek/skip semantic but I would still want the `IReadOnlyList<T>` interface to be implemented to have a more generic way to access the data of that stack.

**Comment by michaelstaib** [comment]
> Isn't the fact that Stack<T> allows random access internally just an implementation detail?

Yes it is an implementation detail since the stack could also have been implemented with a linked list underneath.

But I want to be able to peek ahead without popping items yet. Like a look ahead on the stack. With `ImmutableStack` I could do this for instance. But with an indexer I could do this more efficiently. You could also argue that a stack does not even has to be iterated over. Since If you look at the data structure it would suffice to have Push/Pop/Peek functionality. 

IReadOnlyList implements IReadOnlyCollection. IReadOnlyCollection already brings in a Count and IEnumerable. The one thing we would add with the IReadOnlyList is a readonly indexer.

**Comment by scalablecory** [comment]
Is there something that prevents you simply using `List<T>` as a stack?

**Comment by AlgorithmsAreCool** [comment]
Idk, i actually like this idea. I can imagine scenarios where I need to peek the 2nd or nth item in the stack without mutating it. Why not exploit the array based implementation?

**Comment by khellang** [comment]
> Why not exploit the array based implementation?

Because it's just that - an implementation detail. Exposing random access in the public API of a stack seems... wrong.

I get the point that the implementation probably won't change, but by implementing this proposal, you've ruled out that option completely, unless the new implementation also can provide random access, and even then, the characteristics could change, i.e. O(1) -> O(N)).

**Comment by AlgorithmsAreCool** [comment]
@khellang 

I'm having difficulty imagining another implementation that precludes efficient (O(1)) random reads, but doesn't damage core Pop/Peek/Push performance. (And if there is one, you could just make it `StackSlim<T>` 😈)

This pragmatic functionality that is precluded by principle. 

I don't mean to sound excessively negative and I'm not saying you are being unreasonable, it's just a bummer.

/negativity

**Comment by GSPP** [comment]
Related: https://github.com/dotnet/corefx/issues/4961
